### PR TITLE
add proptests for invalid forms of input

### PR DESCRIPTION
**Pull request by yaahc**
part of #335 

This implements the set of tests proposed by @teor2345 in https://github.com/ZcashFoundation/zebra/pull/549#pullrequestreview-437981450

### CODE REVIEW

**Comment by dconnolly** [comment]
OK what I think these are testing, is that for any possible representative EquihashSolution except for the 415000 block one, they will not validate for that block info, which is ok but not exactly what @teor2345 was recommending if I understand correctly.

I think based on @teor2345's comment on the other PR, what we'd like to do for proptesting this is for every valid EquihashSolution, we check `is_valid`, then we corrupt the nonce/header/solution, and then check that `is_valid` fails.

**Comment by yaahc** [comment]
> I think based on @teor2345's comment on the other PR, what we'd like to do for proptesting this is for every valid EquihashSolution, we check `is_valid`, then we corrupt the nonce/header/solution, and then check that `is_valid` fails.

Would this require implementing the logic to generate valid equihash solutions and accompanying headers or do we already have that ability somewhere?

**Comment by dconnolly** [comment]
> > I think based on @teor2345's comment on the other PR, what we'd like to do for proptesting this is for every valid EquihashSolution, we check `is_valid`, then we corrupt the nonce/header/solution, and then check that `is_valid` fails.
> 
> Would this require implementing the logic to generate valid equihash solutions and accompanying headers or do we already have that ability somewhere?

We have strategies to generate randomized `BlockHeader`s but the `EquihashSolution`s are just random bytes right now:

https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/block/tests.rs

**Review by dconnolly — COMMENTED** [review]
OK what I think these are testing, is that for any possible representative EquihashSolution except for the 415000 block one, they will not validate for that block info, which is ok but not exactly what @teor2345 was recommending if I understand correctly.

I think based on @teor2345's comment on the other PR, what we'd like to do for proptesting this is for every valid EquihashSolution, we check `is_valid`, then we corrupt the nonce/header/solution, and then check that `is_valid` fails.

**Comment by yaahc** [comment]
> I think based on @teor2345's comment on the other PR, what we'd like to do for proptesting this is for every valid EquihashSolution, we check `is_valid`, then we corrupt the nonce/header/solution, and then check that `is_valid` fails.

The part of this I had trouble with was how to do the corruption step. I figured doing some arbitrary bit manipulation was going to be insufficient so I wanted to use proptest to handle generating the piece of the struct I was "corrupting" by generating a random one that isn't equal to the original. I will look into making the valid block an input rather than hardcoded to `415000`, then at least we can pass in each of the blocks we have hard coded in `zebra-test`.

**Comment by teor2345** [comment]
> then at least we can pass in each of the blocks we have hard coded in zebra-test.

It would be nice to have an array of "all the blocks in zebra-test", so new blocks automatically get added to all the tests.

For example, I'd like to add block 2, so we can test checkpoint chaining from 0 to 2.

**Comment by hdevalence** [comment]
Having the bytes of the first few blocks would be very useful also because the coinbase height is badly encoded and the first (I think) 16-20 blocks are all special cased. So testing that part of the encoding would be useful.

**Comment by hdevalence** [comment]
I think it might be better to get a `BlockHeader`-based Equihash API first, since the testing we'll want to do is downstream of the invariants provided by the API, and the API we'll eventually want is differently shaped than the one we have.

**Comment by yaahc** [comment]
@hdevalence should I do that as a separate PR or is it okay if I roll that change up with this PR?

**Comment by hdevalence** [comment]
Either way is fine, but I think it would be a waste of effort to keep getting all the tests to compile while changing the API (rather than getting the API right and then doing the tests).  That's the only reason that I can think of for doing two PRs.  If you want to roll it up with this one I think it's fine for the build to be broken in the meantime.

**Comment by yaahc** [comment]
@hdevalence okay I started out by just not exporting the `EquihashSolution` based is_valid API and added a pub version of that API for `BlockHeader` that serializes itself internally.

File: zebra-chain/src/block/header.rs
Comment (hdevalence) [inline zebra-chain/src/block/header.rs]: This should probably take the `n` and `k` from some kind of consensus-parameters object right?

Comment (yaahc) [reply zebra-chain/src/block/header.rs]: absolutely, do you want me to open an issue for that?

Comment (hdevalence) [reply zebra-chain/src/block/header.rs]: Yeah, I think that we should think about how we want to represent consensus parameters, maybe in an object in `zebra-consensus` or something?

Comment (yaahc) [reply zebra-chain/src/block/header.rs]: https://github.com/ZcashFoundation/zebra/issues/565

File: zebra-chain/src/equihash_solution.rs
Comment (teor2345) [inline zebra-chain/src/equihash_solution.rs]: Suggestion: Please specify explicit sizes or types here. If we don't have a specific field or type, let's make the size into a constant.

```suggestion
    const INPUT_LENGTH: usize = 108;
    pub(crate) fn is_valid(&self, input: [u8; INPUT_LENGTH], nonce: [u8; 32]) -> bool {
```

File: zebra-chain/src/block/header.rs
Comment (teor2345) [inline zebra-chain/src/block/header.rs]: Issue: the equihash is only over the header bytes up to the nonce.

```suggestion
        solution.is_valid(input.as_slice()[0..EquihashSolution::INPUT_LENGTH], nonce)
```

File: zebra-chain/src/equihash_solution.rs
Comment (teor2345) [inline zebra-chain/src/equihash_solution.rs]: Issue: `BlockHeader.is_valid()` has no test coverage.

Suggestion: use the `BlockHeader`-based API in the tests, so we have test coverage of the low-level and high-level APIs.

File: zebra-chain/src/block/header.rs
Comment (teor2345) [inline zebra-chain/src/block/header.rs]: Suggestion: Let's return `Result<(), Error>` here instead?

That would integrate better with the existing Block validation code, which uses `?` on the validation functions it calls.

Comment (teor2345) [inline zebra-chain/src/block/header.rs]: Issue: `BlockHeader.is_valid()` is easily confused with validation for all the fields in the header.

Suggestion: Let's call it `is_equihash_solution_valid()`, or something similar.

File: zebra-chain/src/equihash_solution.rs
Comment (teor2345) [inline zebra-chain/src/equihash_solution.rs:151]: It's great to see more testing!

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for making more tests, and for making an interface that's easier to use.

Issues: The new `BlockHeader`-based function uses the wrong input fields, it doesn't have test coverage, and it has a confusing name.

I also made some minor suggestions about naming and the private API.

**Comment by yaahc** [comment]
blocked by https://github.com/ZcashFoundation/librustzcash/pull/1

File: zebra-chain/Cargo.toml
Comment (yaahc) [inline zebra-chain/Cargo.toml]: This change will be reverted before this PR is merged but is needed for it to compile while https://github.com/ZcashFoundation/librustzcash/pull/1 is unmerged

File: zebra-chain/src/equihash_solution.rs
Comment (yaahc) [reply zebra-chain/src/equihash_solution.rs]: I don't think it's worth trying to make `input` an array over a slice, unless we never work with slices of data we will always have to do a conversion from a slice to an array which can only be done through `try_from`, which is already a runtime check. On top of that `try_from` is only implemented for arrays up to 32 elements in size, once we get const generics this will be better but for now I think its preferable to just have an assertion to validate the `INPUT_LENGTH` because either way this code only gets called in one place where we control the construction of `input`, and in all other cases its just test code and we'd have to do the runtime conversion anyways.

Comment (yaahc) [reply zebra-chain/src/equihash_solution.rs]: Also the subsequent request to just do the tests with the `is_equihash_solution_valid` inspired me to just remove this function entirely so I think this is a moot point now

**Review by teor2345 — DISMISSED** [review]
Looks great!

Let's merge the librustzcash changes, tweak this PR's imports, and merge.
(I don't need to re-review the import tweak.)

**Review by teor2345 — DISMISSED** [review]
Still good :-)

I'll let you squash and merge however you'd like.

File: zebra-chain/Cargo.toml
Comment (teor2345) [reply zebra-chain/Cargo.toml]: I merged ZcashFoundation/librustzcash#1.

Please revert this change, and then you can admin-merge, if you'd like :-)

**Comment by teor2345** [comment]
Please also revert the cargo changes :-)

**Pull request merged**
